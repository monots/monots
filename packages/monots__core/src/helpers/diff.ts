import chalk from 'chalk';
import chalkTemplate from 'chalk-template';
import isEqual from 'fast-deep-equal';
import { diff } from 'jest-diff';
import * as path from 'node:path';

import { FatalError } from './errors.js';

const diffOptions = {
  contextLines: 1,
  expand: false,
  aAnnotation: 'Current',
  aColor: chalk.red,
  bAnnotation: 'Expected',
  bColor: chalk.green,
  includeChangeCounts: true,
};

/**
 * Sort the keys alphabetically to produce consistent comparisons.
 */
function orderOutputKeys(output: Record<string, unknown>) {
  return Object.keys(output)
    .sort()
    .map((name) => path.relative(process.cwd(), name));
}

interface CompareOutputProps {
  actual: Record<string, unknown>;
  expected: Record<string, unknown>;
  name: string;
}

/**
 * Check that the actual output and the expected output are identical. When
 * content has changed it will throw an error with a descriptive diff.
 */
export function compareOutput(props: CompareOutputProps) {
  const { actual, expected, name } = props;
  const actualKeys = orderOutputKeys(actual);
  const expectedKeys = orderOutputKeys(expected);

  if (!isEqual(actualKeys, expectedKeys)) {
    return [
      new FatalError(
        chalkTemplate`\n{yellow The package.json file has unexpected keys.}\n\n${
          diff(actualKeys, expectedKeys, diffOptions) || ''
        }`,
        name,
      ),
    ];
  }

  const errorMessages: string[] = [];

  for (const [name, actualContents] of Object.entries(actual)) {
    const expectedContents = expected[name];
    const relativeName = path.relative(process.cwd(), name);

    if (isEqual(actualContents, expectedContents)) {
      continue;
    }

    errorMessages.push(
      chalkTemplate`{grey ${relativeName}}\n${diff(actualContents, expected[name], diffOptions)}`,
    );
  }

  if (errorMessages.length > 0) {
    return [
      new FatalError(
        chalkTemplate`\n{bold.yellow The actual content differs from expected content.}\n\n${errorMessages.join(
          '\n\n',
        )}\n`,
        name,
      ),
    ];
  }

  return [];
}
