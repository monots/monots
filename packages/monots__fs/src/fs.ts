import type { GlobProps } from '@monots/glob';
import { getPath, glob } from '@monots/glob';
import type { IPromisesAPI } from 'memfs/lib/promises.js';
import type { DirectoryJSON } from 'memfs/lib/volume.js';
import { Volume } from 'memfs/lib/volume.js';
import * as fsPromises from 'node:fs/promises';
import path from 'node:path';

import type { AnyFunction, Method, NativeFsPromises as NativeFsPromises } from './types';
import { METHODS } from './types';

type FileSystem = 'native' | 'memory';

/**
 * Create a file system for use in monots.
 */
export class Fs implements NativeFsPromises {
  #volume: Volume;
  #promises: IPromisesAPI;
  #backend: FileSystem;

  get backend(): FileSystem {
    return this.#backend;
  }

  get volume(): Volume {
    return this.#volume;
  }

  constructor(props: FsProps = {}) {
    this.#backend = props.backend ?? 'native';
    this.#volume = new Volume({});
    this.#promises = this.#volume.promises;

    for (const method of METHODS) {
      this[method] = this.#getMethod(method);
    }
  }

  /**
   * Populate the memory file system from the provide path to the path on the
   * memory file system.
   */
  async populateFromFileSystem(props: PopulateMemory) {
    const { from, target = from, ...rest } = props;
    const promises: Array<Promise<readonly [file: string, contents: string]>> = [];
    const preserve = from === target;
    const targetPath = getPath(target);
    const iterator = glob({
      ...rest,
      cwd: from,
      trailingSlash: true,
      includeDirectories: false,
      dot: props.dot ?? true,
      caseInsensitive: props.caseInsensitive ?? true,
    });

    for await (const file of iterator) {
      const promise = fsPromises.readFile(file.absolute, 'utf8').then((contents) => {
        return [preserve ? file.absolute : path.join(targetPath, file.relative), contents] as const;
      });

      promises.push(promise);
    }

    const entries = await Promise.all(promises);
    this.#volume.fromJSON(Object.fromEntries(entries));

    return this;
  }

  /**
   * Populate the memory file system with the provided json.
   */
  populateFromJson(json: DirectoryJSON) {
    this.#volume.fromJSON(json);
    return this;
  }

  /**
   * Commit the files to the file system.
   */
  async commitToNative() {}

  /**
   * Load the current JSON stored in memory.
   */
  loadJsonFromMemory() {
    this.#volume.createReadStream;
    return this.#volume.toJSON();
  }

  /**
   * Switch to use the native node file system.
   */
  useNative() {
    this.#backend = 'native';
    return this;
  }

  useMemory() {
    this.#backend = 'memory';
    return this;
  }

  #getMethod = (method: Method): AnyFunction => {
    if (this.#backend === 'memory' && ['lutimes', 'opendir', 'watch', 'cp'].includes(method)) {
      throw new Error(`"${method}" is not supported in memory file system`);
    }

    return (...args: any[]) => {
      return this.#backend === 'native'
        ? (fsPromises as any)[method](...args)
        : (this.#promises as any)[method](...args);
    };
  };
}

interface FsProps {
  /**
   * @default 'native'
   */
  backend?: FileSystem;
}

interface PopulateMemory
  extends Pick<GlobProps, 'extensions' | 'exclude' | 'include' | 'dot' | 'caseInsensitive'> {
  /**
   * The absolute path to the file system to populate from.
   */
  from: URL | string;

  /**
   * The absolute path to target. If left undefined then this defaults to using
   * same location as from.
   */
  target?: URL | string;
}

export interface Fs extends NativeFsPromises {}
export const fs = new Fs();
export const {
  access,
  appendFile,
  chmod,
  chown,
  copyFile,
  cp,
  lchown,
  link,
  lstat,
  lutimes,
  mkdir,
  mkdtemp,
  open,
  opendir,
  readdir,
  readFile,
  readlink,
  realpath,
  rename,
  rm,
  rmdir,
  stat,
  symlink,
  truncate,
  unlink,
  utimes,
  watch,
  writeFile,
} = fs;
