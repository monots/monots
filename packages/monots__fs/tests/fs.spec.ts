import { fs } from '@monots/fs';
import { test } from 'vitest';
import { diffJson, diffArrays } from 'diff';
import {} from '@monots/test';

test('can switch between native and memory', async ({ expect }) => {
  expect(fs.backend).toBe('native');
  fs.useMemory();
  expect(fs.backend).toBe('memory');
});

test.only('populate from file system and apply changes back', async ({ expect }) => {
  await fs.useMemory().populateFromFileSystem({
    from: new URL('fixtures', import.meta.url),
    target: '/custom/path',
  });
  const snapshot = fs.loadJsonFromMemory();
  expect(snapshot).toMatchSnapshot();
  await fs.appendFile(new URL('/custom/path/file.md', import.meta.url), 'hello world');
  await fs.writeFile(new URL('/custom/path/files.md', import.meta.url), 'hello world');
  await fs.rm(new URL('/custom/path/.file', import.meta.url));
  expect(fs.loadJsonFromMemory()).toMatchSnapshot();
});
