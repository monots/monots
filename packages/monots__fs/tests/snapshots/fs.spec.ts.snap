// Vitest Snapshot v1

exports[`populate from file system 1`] = `
{
  "/custom/path/.directory/file.md": "",
  "/custom/path/.file": "hello my sister!
",
  "/custom/path/file.md": "with contents
",
  "/custom/path/first/file.md": "the first time i saw your face, i wept for you
",
  "/custom/path/first/nested/directory/file.json": "{}
",
  "/custom/path/first/nested/directory/file.md": "",
  "/custom/path/first/nested/file.md": "you are my fire
",
  "/custom/path/second/file.md": "you got this!
",
  "/custom/path/second/nested/directory/file.md": "",
  "/custom/path/second/nested/file.md": "why??!!!
",
}
`;

exports[`populate from file system and apply changes back 1`] = `
{
  "/custom/path/.directory/file.md": "",
  "/custom/path/.file": "hello my sister!
",
  "/custom/path/file.md": "with contents
",
  "/custom/path/first/file.md": "the first time i saw your face, i wept for you
",
  "/custom/path/first/nested/directory/file.json": "{}
",
  "/custom/path/first/nested/directory/file.md": "",
  "/custom/path/first/nested/file.md": "you are my fire
",
  "/custom/path/second/file.md": "you got this!
",
  "/custom/path/second/nested/directory/file.md": "",
  "/custom/path/second/nested/file.md": "why??!!!
",
}
`;

exports[`populate from file system and apply changes back 2`] = `
{
  "/custom/path/.directory/file.md": "",
  "/custom/path/file.md": "with contents
hello world",
  "/custom/path/files.md": "hello world",
  "/custom/path/first/file.md": "the first time i saw your face, i wept for you
",
  "/custom/path/first/nested/directory/file.json": "{}
",
  "/custom/path/first/nested/directory/file.md": "",
  "/custom/path/first/nested/file.md": "you are my fire
",
  "/custom/path/second/file.md": "you got this!
",
  "/custom/path/second/nested/directory/file.md": "",
  "/custom/path/second/nested/file.md": "why??!!!
",
}
`;
